File: Orchestrators/DotNet/Workflows/WorkflowOrchestrator.cs
using AITrove.Agents.Interfaces;
using AITrove.Context;
using Microsoft.Extensions.Logging;

namespace AITrove.Workflows;

/// <summary>
/// Fan-out + synthesize orchestrator.
///
/// Pattern:
///   1. Discover eligible agents via IAgent.CanHandle(ctx)
///   2. Run them all concurrently with Task.WhenAll
///   3. Collect successful outputs into AgentContext
///   4. Ask Claude to synthesize a final answer
///
/// Failures in individual agents are logged and skipped — they never
/// block the workflow unless ALL agents fail.
/// </summary>
public sealed class WorkflowOrchestrator(
    IEnumerable<IAgent> agents,
    IAnthropicMessageClient anthropic,
    ILogger<WorkflowOrchestrator> logger)
{
    private readonly IReadOnlyList<IAgent> _agents = [.. agents];

    public async Task<string> RunAsync(AgentContext ctx, CancellationToken ct = default)
    {
        var eligible = _agents.Where(a => a.CanHandle(ctx)).ToList();
        logger.LogInformation("Workflow starting. Eligible agents: {Count}", eligible.Count);

        if (eligible.Count == 0)
        {
            logger.LogWarning("No eligible agents found. Passing user message directly to Claude.");
            return await SynthesizeAsync(ctx, ct);
        }

        // Fan out — all eligible agents run concurrently
        var results = await Task.WhenAll(eligible.Select(a => RunAgentSafe(a, ctx, ct)));

        // Collect successful outputs
        foreach (var r in results.Where(r => r.Success))
            ctx.AddAgentResult(r.AgentName, r.Output);

        logger.LogInformation(
            "{Success}/{Total} agents succeeded",
            results.Count(r => r.Success), results.Length);

        return await SynthesizeAsync(ctx, ct);
    }

    // -------------------------------------------------------------------------

    private async Task<AgentResult> RunAgentSafe(
        IAgent agent, AgentContext ctx, CancellationToken ct)
    {
        try
        {
            logger.LogDebug("Running agent: {Name}", agent.Name);
            var output = await agent.ExecuteAsync(ctx, ct);
            return new AgentResult(agent.Name, output, true);
        }
        catch (Exception ex)
        {
            logger.LogWarning(ex, "Agent {Name} failed — skipping", agent.Name);
            return new AgentResult(agent.Name, string.Empty, false);
        }
    }

    private async Task<string> SynthesizeAsync(AgentContext ctx, CancellationToken ct)
    {
        return await anthropic.SendMessageAsync(
            ctx.ApiKey,
            AnthropicModelIds.ClaudeSonnet4,
            maxTokens: 1024,
            systemPrompt: ctx.SystemPrompt,
            userMessage: ctx.BuildSynthesisPrompt(),
            ct);
    }
}

// ---------------------------------------------------------------------------
public record AgentResult(string AgentName, string Output, bool Success);

File: Orchestrators/Python/aitrove_python/workflow_orchestrator.py
from __future__ import annotations

import asyncio
import logging
from dataclasses import dataclass

from .agent_context import AgentContext
from .anthropic_client import AnthropicMessageClient, AnthropicModelIds
from .agents.base import Agent


@dataclass(slots=True)
class AgentResult:
    agent_name: str
    output: str
    success: bool


class WorkflowOrchestrator:
    def __init__(
        self,
        agents: list[Agent],
        anthropic: AnthropicMessageClient,
        logger: logging.Logger,
    ) -> None:
        self._agents = list(agents)
        self._anthropic = anthropic
        self._logger = logger

    async def run(self, ctx: AgentContext) -> str:
        eligible = [agent for agent in self._agents if agent.can_handle(ctx)]
        self._logger.info("Workflow starting. Eligible agents: %s", len(eligible))

        if not eligible:
            self._logger.warning("No eligible agents found. Passing user message directly to Anthropic.")
            return await self._synthesize(ctx)

        results = await asyncio.gather(*(self._run_agent_safe(agent, ctx) for agent in eligible))

        for result in results:
            if result.success:
                ctx.add_agent_result(result.agent_name, result.output)

        self._logger.info("%s/%s agents succeeded", sum(1 for result in results if result.success), len(results))
        return await self._synthesize(ctx)

    async def _run_agent_safe(self, agent: Agent, ctx: AgentContext) -> AgentResult:
        try:
            self._logger.debug("Running agent: %s", agent.name)
            output = await agent.execute(ctx)
            return AgentResult(agent.name, output, True)
        except Exception:
            self._logger.exception("Agent %s failed — skipping", agent.name)
            return AgentResult(agent.name, "", False)

    async def _synthesize(self, ctx: AgentContext) -> str:
        return await self._anthropic.send_message(
            api_key=ctx.api_key,
            requested_model=AnthropicModelIds.CLAUDE_SONNET_4,
            max_tokens=1200,
            system_prompt=ctx.system_prompt,
            user_message=ctx.build_synthesis_prompt(),
        )
